runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java

```diff
@@ -281,52 +170,200 @@ public Object applyUpdateUnsafe(SMREntry entry, boolean isOptimistic) {
-            // optimistic undo log, and mark that the object is
-            // no longer optimistically undoable if we
-            // are in optimistic mode.
-            if (isOptimistic) {
-                optimisticUndoLog.clear();
-            } else {
-                undoLog.clear();
+
+        throw new NoRollbackException();
+    }
+
+    /** Move the pointer for this object (effectively, forcefuly
+     * change the version of this object without playing
+     * any updates).
+     * @param globalAddress     The global address to set the pointer to
+     */
+    public void seek(long globalAddress) {
+        smrStream.seek(globalAddress);
+    }
+
+    protected void syncStreamUnsafe(ISMRStream stream, long timestamp) {
+        log.trace("Sync[{}] {}", this, (timestamp == Address.OPTIMISTIC)
+                                            ? "Optimistic" : "to " + timestamp);
+        long syncTo = (timestamp == Address.OPTIMISTIC) ? Address.MAX : timestamp;
+        stream.remainingUpTo(syncTo)
+                .stream()
+                .forEachOrdered(entry -> {
+                    try {
+                        Object res = applyUpdateUnsafe(entry);
+                        if (timestamp == Address.OPTIMISTIC) {
+                            entry.setUpcallResult(res);
+                        }
+                        else if (pendingUpcalls.contains(entry.getEntry().getGlobalAddress())) {
+                            log.debug("Sync[{}] Upcall Result {}", entry.getEntry().getGlobalAddress());
+                            upcallResults.put(entry.getEntry().getGlobalAddress(), res == null ?
+                                    NullValue.NULL_VALUE : res);
+                            pendingUpcalls.remove(entry.getEntry().getGlobalAddress());
+                        }
+                        entry.setUpcallResult(res);
+                    } catch (Exception e) {
+                        log.error("Sync[{}] Error: Couldn't execute upcall due to {}", this, e);
+                        throw new RuntimeException(e);
+                    }
+                });
+    }
+    /** Update the object. Ensure that you have the write lock before calling
+     * this function...
+     * @param timestamp         The timestamp to update the object to.
+     */
+    public void syncObjectUnsafe(long timestamp) {
+        // If there is an optimistic stream attached,
+        // and it belongs to this thread use that
+        if (optimisticStream != null &&
+                optimisticStream.isStreamForThisTransaction()) {
+            // If there are no updates, ensure we are at the right snapshot
+            if (optimisticStream.pos() == Address.NEVER_READ) {
+                final WriteSetSMRStream currentOptimisticStream =
+                        optimisticStream;
+                // It's necessary to roll back optimistic updates before
+                // doing a sync of the regular log
+                optimisticRollbackUnsafe();
```

Comment:
why is this here? we just checked above that (a) we are the owner of the proxy, and (b) that there are not optimistic updates applied.

also, if there were any updates, it would make me a bit nervous to undo them and then redo over a different version?